presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/ReentrantExecutor.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.metastore.cache;
+
+import java.util.concurrent.Executor;
+
+import static java.util.Objects.requireNonNull;
+
+public class ReentrantExecutor
+        implements Executor
+{
+    private final ThreadLocal<Boolean> executorThreadMarkers = ThreadLocal.withInitial(() -> false);
```

Comment:
It's tri-state effectively (true, false, but could be unset).
You can make it two-state by using `Object MARKER` and only `set, remove` methods.